README.md

```diff
@@ -1 +1,180 @@
+
+`DbLinq` static class defined in `Starcounter.Linq` namespace should be used as starting point when you want to build LINQ queries. It contains two important methods: `Objects<T>` and `CompileQuery<T>` (many overloaded ones) which are used for two different approaches. For example:
+
+```csharp
+// ad-hoc request
+var johns = DbLinq.Objects<Person>().Where(p => p.Name == "John");
+
+// request using compiled query
+var peopleByNameQuery = DbLinq.CompileQuery((string name) =>
+                        DbLinq.Objects<Person>().Where(p => p.Name == name));
+var jennifers = peopleByNameQuery("Jennifer");
+```
+
+## Ad-hoc requests
+
+Using this approach you build a LINQ expression and obtain data as usual when an application needs it.
+
+`DbLinq.Objects<T>()` returns a data context `Queryable<T>` which can be used for obtaining data.
+
+We recommend to use it only in cases when a request is expected to be executed infrequently. The reason is that it is very expensive to translate LINQ expression to SQL. Using this approach such translating will be done every calling.
+
+Example:
+
+```csharp
+// this method is used quite rare, so we can use ad-hoc requests here
+void Handle(Input.DeleteGroupTrigger action)
+{
+    Db.Transact(() =>
+    {
+        var surfaces = DbLinq.Objects<WebTemplate>().Where(x => x.WebTemplateGroup == this.Data);
+
+        foreach (WebTemplate surface in surfaces)
+        {
+            surface.WebTemplateGroup = null;
+        }
+        this.Data.Delete();
+    });
+}
+```
+
+## Compiled Query
+
+Compiled Query lets you build a LINQ expression with translated SQL once and use it many times.
+
+`DbLinq.CompileQuery<T>` returns a delegate `Func<IEnumerable<T>>` which can be invoked to execute the compiled query. It has many overloads and it supports passing parameters.
+
+We recommend to use in all cases when a request is executed quite frequently.
+
+Example:
+
+```csharp
+partial class SurfacePage : Json
+{
+    private static readonly Func<IEnumerable<WebTemplate>> SurfaceGroupsQuery = DbLinq.CompileQuery(() =>
+            DbLinq.Objects<WebTemplateGroup>().Where(x => !x.Deleted));
+
+    private static readonly Func<WebTemplate, IEnumerable<WebUrl>> RoutesQuery =
+        DbLinq.CompileQuery((WebTemplate surface) =>
+            DbLinq.Objects<WebUrl>()
+                .Where(x => !x.Deleted && !x.Hidden)
+                .Where(x => x.Template == surface)
+                .OrderBy(x => x.SortNumber));
+    /*
+    ...
+    */
+
+    // this method is called quite frequently so we should use compiled queries
+    public void RefreshData()
+    {
+        this.SurfaceGroups.Data = SurfaceGroupsQuery();
+        this.Routes.Data = RoutesQuery(this.Data);
+    }
+}
+```
+
+Please note, that Compiled Query has some additional restrictions since it represents a pre-translated SQL and should support passing parameters.

```

Comment:
> Compiled queries are more restricted than ad-hoc requests since it represents a pre-translated SQL and should support passing parameters. Read more about the restrictions below.